### PR TITLE
Add torch.aten.batch_norm Linalg lowering support

### PR DESCRIPTION
**Pull request by cathyzhyi**
1. Added a simplified version of torch.aten.batch_norm which only handles
inference and assumes the weight, bias, running_mean, running_var are not
None.

2. Removed the primitive types check in verifyLinalgCompatibleTypes check
since now we have proper type converter to handle torch types conversion.
The checks for RankedTensorType is kept because the type converter
doesn't guarantee the converted builtin tensor type is ranked. A
separate verification pass to verify the invariant expected by later
passes will need to be added before those can be removed as well.

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 1]
```diff
@@ -42,30 +42,128 @@ using namespace mlir::NPCOMP::Torch;
...
-              tensor.toBuiltinTensor().dyn_cast_or_null<RankedTensorType>()) {
-        if (BaseMemRefType::isValidElementType(rankedTensor.getElementType()))
-          return true;
-      }
-    }
-    if (type.isa<mlir::FloatType, IntegerType, IndexType>())
-      return true;
-    return false;
+    auto tensor = type.dyn_cast<ValueTensorType>();
+    return !tensor || tensor.toBuiltinTensor().isa<RankedTensorType>();
```
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:50]: toBuiltinTensor can return null. So use dyn_cast_or_null (there is no isa_or_null) and rely on implicit boolean check.

File: frontends/pytorch/e2e_testing/torchscript/basic.py [hunk 2]
```diff
@@ -101,3 +101,22 @@ def forward(self, x):
 @register_test_case(module_factory=lambda: ReluModule())
 def ReluModule_basic(module, tu: TestUtils):
     module.forward(tu.rand(4, 2) - 0.5)
+
+class BatchNormModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.bn2d = torch.nn.BatchNorm2d(2)
+        self.bn2d.eval()
+        self.bn2d.running_mean = torch.tensor([0.5, 0.4])
```
Comment (silvasean) [inline frontends/pytorch/e2e_testing/torchscript/basic.py:110]: can you remove running_mean/var for now and add a TODO explaining the work remaining to be done?

File: frontends/pytorch/e2e_testing/torchscript/basic.py [hunk 3]
```diff
@@ -101,3 +101,22 @@ def forward(self, x):
 @register_test_case(module_factory=lambda: ReluModule())
 def ReluModule_basic(module, tu: TestUtils):
     module.forward(tu.rand(4, 2) - 0.5)
+
+class BatchNormModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.bn2d = torch.nn.BatchNorm2d(2)
```
Comment (silvasean) [inline frontends/pytorch/e2e_testing/torchscript/basic.py:108]: Since you implemented support for many cases, can you also add BatchNorm1d and BatchNorm3d to verify the extreme cases of NC and NCDHW?

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 4]
```diff
@@ -42,30 +42,128 @@ using namespace mlir::NPCOMP::Torch;
...
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, input.getType(),
+                ValueRange{input, weight, bias, runningMean, runningVar}, input,
+                /*indexingMaps=*/indexingMaps,
+                /*iteratorTypes=*/iteratorTypes,
+                [&](OpBuilder &b, Location loc, ValueRange args) {
+                  Value input = args[0], weight = args[1], bias = args[2],
+                        mean = args[3], var = args[4];
+                  // (input - mean) / sqrt(var + eps) * weight + bias
```
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:148]: add more parentheses to make it clearer what the `* weight` is applied to.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 5]
```diff
@@ -42,30 +42,128 @@ using namespace mlir::NPCOMP::Torch;
...
+                  Value inputSubMean = b.create<SubFOp>(loc, input, mean);
+                  // The eps is always f64.
+                  Value exEps = b.create<FPTruncOp>(loc, var.getType(), eps);
```
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:151]: what is "ex" in "exEps"? Can you rename it?

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 6]
```diff
@@ -42,30 +42,128 @@ using namespace mlir::NPCOMP::Torch;
...
+                  Value varPlusEps = b.create<AddFOp>(loc, var, exEps);
+                  Value rSTD = b.create<math::RsqrtOp>(loc, varPlusEps);
+                  Value temp = b.create<MulFOp>(loc, inputSubMean, rSTD);
+                  Value timeWeight = b.create<MulFOp>(loc, temp, weight);
```
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:155]: nit: time -> times

**Review by silvasean — CHANGES_REQUESTED** [review]

File: frontends/pytorch/e2e_testing/torchscript/basic.py (hunk 2 above)
Comment (cathyzhyi) [reply frontends/pytorch/e2e_testing/torchscript/basic.py:110]: If the `running_mean/var` is not given the input will always be the same as the output. The Batchnorm calculation would not really be fully tested.  Maybe I can add comments to explain why the fake data is added for testing purpose?

Comment (silvasean) [reply frontends/pytorch/e2e_testing/torchscript/basic.py:110]: Oh, I see, you are initializing the BatchNorm2d instance. For some reason I thought it was `self.running_mean`. Sorry!

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp (hunk 5 above)
Comment (cathyzhyi) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp]: I was somehow thinking about extending but this is actually truncating. Changed the name to truncatedEps.

**Review by silvasean — APPROVED** [review]
